src/main/java/edu/wpi/grip/core/serialization/SocketConverter.java

```diff
@@ -0,0 +1,170 @@
+            }
+
+            // Look up the socket using the saved indexes.  Serializing sockets this way makes it so different things
+            // (such as connections) can reference sockets in the pipeline.
+            Socket socket;
+            if (reader.getAttribute(STEP_ATTRIBUTE) != null) {
+                final int stepIndex = Integer.valueOf(reader.getAttribute(STEP_ATTRIBUTE));
+                final int socketIndex = Integer.valueOf(reader.getAttribute(SOCKET_ATTRIBUTE));
+
+                final Step step = pipeline.getSteps().get(stepIndex);
+                socket = direction == Socket.Direction.INPUT ?
+                        step.getInputSockets()[socketIndex] : step.getOutputSockets()[socketIndex];
+            } else if (reader.getAttribute(SOURCE_ATTRIBUTE) != null) {
+                final int sourceIndex = Integer.valueOf(reader.getAttribute(SOURCE_ATTRIBUTE));
+                final int socketIndex = Integer.valueOf(reader.getAttribute(SOCKET_ATTRIBUTE));
+
+                final Source source = pipeline.getSources().get(sourceIndex);
+                socket = source.getOutputSockets()[socketIndex];
+            } else {
+                throw new ConversionException("Sockets must have either a step or source attribute");
+            }
+
+            if (socket.getDirection() == Socket.Direction.OUTPUT) {
+                ((OutputSocket) socket).setPublished(Boolean.valueOf(reader.getAttribute(PUBLISHED_ATTRIBUTE)));
+                ((OutputSocket) socket).setPreviewed(Boolean.valueOf(reader.getAttribute(PREVIEWED_ATTRIBUTE)));
+            }
+
+            if (reader.hasMoreChildren()) {
+                reader.moveDown();
+                if (reader.getNodeName().equals("value")) {
+                    socket.setValue(context.convertAnother(socket, getDeserializedType(socket)));
+                } else {
+                    throw new IllegalArgumentException("Unexpected node in socket: " + reader.getNodeName());
+                }
+                reader.moveUp();
+            }
+
+            reader.moveUp();
+
+            return socket;
+        } catch (Exception e) {
+            throw new ConversionException("Error deserializing socket", e);
+        }
+    }
+
+    /**
+     * @param socket
+     * @return The type to convert a serialized socket value into when deserializing.  This can't always be the type in
+     * the socket, since sockets very often hold interfaces, and any number of classes may implement that interface.
+     */
+    private Class<?> getDeserializedType(Socket socket) {
+        final Class<?> socketType = socket.getSocketHint().getType();
+
+        if (!socketType.isInterface() && !Modifier.isAbstract(socketType.getModifiers())) {
+            return socketType;
+        } else if (socketType.equals(List.class)) {
+            return ArrayList.class;
+        } else if (socketType.equals(Number.class)) {
+            return Double.class;
+        }
```

Comment:
Nope.  If a future developer creates a socket hint with type `List`, it shouldn't be unexpected that that socket gets assigned a value of type `ArrayList`, since `ArrayList` implements `List`.

Also, it doesn't convert socket values.  If you read the comments, it converts the *serialized* data into an ArrayList.  That means that if you have a socket value like this:

    <value>
        <double>1</double>
        <double>2</double>
        <double>3</double>
    </value>

... that we should store 1, 2, and three in an ArrayList.  There is literally no way to avoid choosing a concrete list class to use.